### DISCUSSION TITLE
SofaPython3 controller : pause the animation after a given number of simulation steps

### DISCUSSION THREAD

**Discussion by jducrocq007** [sofa-framework/sofa#4372] [SofaPython3] [answered]
Hello everyone :)

I hope that you're all ok!

So actually in my simulation, I have a SofaPython 3 controller which is working nicely, which changes the states of my mechanical objects during runtime.

However, I would like to automatically pause the animation after a given number of simulation steps (ex. : 10), WITHIN the controller.

Thanks to this post in the old Sofa forum, I found this : 
[Sofa old forum : stop animation](https://www.sofa-framework.org/community/forum/topic/stop-animation-function-in-sofa/)
which explains how to STOP the animation. 
It works, but since the simulation is stopped, it resets the DoFs of my system, and that's not what i want!

Therefore, I've been browsing about how to pause the simulation on SofaAPI and forums, i did find some information but there is a piece of code I can't manage to infer...

These are the 2 most relatable links I have:
[sofa pause the animation on the API](https://www.sofa-framework.org/community/doc/programming-with-sofa/api-overview/pause-the-animation/)
[Other explanations about pausing](https://www.sofa-framework.org/api/master/sofa/html/namespacesofa_1_1component_1_1misc.html#ad18fdb65e56306821f273ad9a0ec0bf5)

However, I'm failing to find how to write this "pause" function as a Python code...

Here are my unsuccessful trials:

```
self.node.getRootContext().pause()
self.node.pause()
Sofa.Simulation.pause()
Sofa.simulation.Simulation.pause()
Sofa.Component.misc.PauseAnimation()
Sofa.misc.PauseAnimation()
self.node.getRootContext().PauseAnimation()
Sofa.Simulation.PauseAnimation()
self.node.getRootContext().Simulation.pause()
```

If one of you guys have any hint on pausing the simulation with SofaPython3 scripts, I'd love a helping hand 🥉 

Best,

Julien Ducrocq

**Comment by ISIR-GuillaumeS** [comment]
Hi !

Any news about your issue ?
Would it be possible to have access to your python script ? I try to do the same kind of things but I don't really know where to start !

Thanks ! :smiley:

**Reply by jducrocq007** [reply, marked as answer]
Heyyy sorry it's a very old thing for me, I needed to have results quickly at that time so I did a very simple workaround like this.... 

```
# called at each begin of animation step
def onAnimateBeginEvent(self, event):
# Animate the scene before Z_iter is reached
if self.i < Z_iter:
	for j in range(Z_len):
		myActusCur[j][2] = myActusCur[j][2] + Zstep_actus[j]
	self.node.actuatorsPos.position = myActusCur
	print(str(self.i + 1) + "/" + str(Z_iter))
	self.i = self.i + 1

# Terminate the script after the number of iterations to reach Zstar is done
elif self.i == Z_iter:
	print("Simulation is over: terminating the animation")

	self.node.getRootContext().animate = False  
	self.i = self.i + 1
```

**Reply by jducrocq007** [reply]
It stops the animation after a decided number of iterations (in my case, Z_iter). And sorry, I did not improve more the script because i am currently starting another project, which has nothing to do with Sofa...
Hope this small answer could help you anyways!

**Reply by ISIR-GuillaumeS** [reply]
Hi ! Thanks for sharing, kind of helping for me !
